state/modelsummaries_test.go

```diff
@@ -0,0 +1,390 @@
+	// User 2 has read access to the shared model
+	_, err = sharedModel.AddUser(state.UserAccessSpec{
+		User:      user2.UserTag(),
+		CreatedBy: owner,
+		Access:    "read",
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	_, err = sharedModel.AddUser(state.UserAccessSpec{
+		User:      user3.UserTag(),
+		CreatedBy: owner,
+		Access:    "admin",
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	return modelUUIDs
+}
+
+func (s *ModelSummariesSuite) modelNamesForUser(c *gc.C, user string) []string {
+	tag := names.NewUserTag(user)
+	isSuper, err := s.State.IsUserSuperuser(tag)
+	c.Assert(err, jc.ErrorIsNil)
+	modelQuery, closer, err := s.State.ModelQueryForUser(tag, isSuper)
+	defer closer()
+	c.Assert(err, jc.ErrorIsNil)
+	var docs []struct {
+		Name string `bson:"name"`
+	}
+	modelQuery.Select(bson.M{"name": 1})
+	err = modelQuery.All(&docs)
+	c.Assert(err, jc.ErrorIsNil)
+	names := make([]string, 0)
+	for _, doc := range docs {
+		names = append(names, doc.Name)
+	}
+	sort.Strings(names)
+	return names
+}
+
+func (s *ModelSummariesSuite) TestModelsForUserAdmin(c *gc.C) {
+	s.Setup4Models(c)
+	names := s.modelNamesForUser(c, s.Model.Owner().Name())
+	// Admin always gets to see all models
+	c.Check(names, gc.DeepEquals, []string{"shared", "testenv", "user1model", "user2model", "user3model"})
+}
+
+func (s *ModelSummariesSuite) TestModelsForSuperuserWithoutAll(c *gc.C) {
+	s.Setup4Models(c)
+	summaries, err := s.State.ModelSummariesForUser(s.Model.Owner(), false)
+	c.Assert(err, jc.ErrorIsNil)
+	names := make([]string, len(summaries))
+	for i, summary := range summaries {
+		names[i] = summary.Name
+	}
+	sort.Strings(names)
+	c.Check(names, gc.DeepEquals, []string{"shared", "testenv"})
+}
+
+func (s *ModelSummariesSuite) TestModelsForSuperuserWithAll(c *gc.C) {
+	s.Setup4Models(c)
+	summaries, err := s.State.ModelSummariesForUser(s.Model.Owner(), true)
+	c.Assert(err, jc.ErrorIsNil)
+	names := make([]string, len(summaries))
+	access := make(map[string]string)
+	for i, summary := range summaries {
+		names[i] = summary.Name
+		access[summary.Name] = string(summary.Access)
+	}
+	sort.Strings(names)
+	c.Check(names, gc.DeepEquals, []string{"shared", "testenv", "user1model", "user2model", "user3model"})
+	c.Check(access, gc.DeepEquals, map[string]string{
+		"shared":     "admin",
+		"testenv":    "admin",
+		"user1model": "",
+		"user2model": "",
+		"user3model": "",
+	})
+}
+
+func (s *ModelSummariesSuite) TestModelsForUser1(c *gc.C) {
+	// User1 is only added to the model they own and the shared model
+	s.Setup4Models(c)
+	names := s.modelNamesForUser(c, "user1write")
+	c.Check(names, gc.DeepEquals, []string{"shared", "user1model"})
+}
+
+func (s *ModelSummariesSuite) TestModelsForUser2(c *gc.C) {
+	// User1 is only added to the model they own and the shared model
+	s.Setup4Models(c)
+	names := s.modelNamesForUser(c, "user2read")
+	c.Check(names, gc.DeepEquals, []string{"shared", "user2model"})
+}
+
+func (s *ModelSummariesSuite) TestModelsForUser3(c *gc.C) {
```

Comment:
Does this test the same as TestModelForUser2? (just for a different user but the same in principal...)

Reply:
User 3 is admin on the shared model, user 2 is read access on the shared model, user 1 is write access.
I updated the comments to indicate as such.